Add selfsigned auth means (#260)
nuts-foundation/nuts-specification: rfc/rfc019-employee-identity-means.md

# RFC019 Employee Identity Authentication Means

|                           |                 |
|:--------------------------|:----------------|
| Nuts foundation           | W.M. Slakhorst  |
| Request for Comments: 019 | Nedap           |
|                           | S. van der Vegt |
|                           | Nedap           |
|                           | April 2023      |

## Employee Identity Authentication Means

### Abstract

The NutsEmployeeIdentity method is an authentication _means_ which allows an organisation to assert the identity of the current user.

> The organisation uses the information of the current user session to present the user with a statement about its identity and asks the user to confirm this.

Reviewer: ```suggestion
The organisation uses the information of the current user session to present the user with a statement about its identity and asks the user to confirm this.
```

I'd say "existing" or "active" user session

After confirmation this bundle of information is signed with the organisation's private key.

This method makes it possible to assert a user identity without the need for a personal authentication means.
This puts the organisation in the role of identity issuer instead of a trusted third party.
A use-case can choose to allow this authentication under certain (low risk) circumstances which still require a person to be authenticated in order to comply with the applicable legislation.

This RFC is an addition to the means listed in [RFC002](./rfc002-authentication-token.md#7-supported-means)

### Status of document

This document is currently in draft.

### Copyright Notice

![](../.gitbook/assets/license.png)

This document is released under the [Attribution-ShareAlike 4.0 International \(CC BY-SA 4.0\) license](https://creativecommons.org/licenses/by-sa/4.0/).

## 1.  Introduction

> Introducing new authentication means in an organizational setting can have quite an impact, even more when mainstream identity providers have not yet adjusted to new standards.

Reviewer: > even more when mainstream identity providers have not yet adjusted to new standards.

How applicable is this? IRMA vs mainstream identity providers? And IRMA isn't a standard?

Either make it more specific (IRMA) or abstract: it's about an authentication means which is under sole control of the user (self-sovereign).

Maybe just use "self-sovereign identity"?

This introduces an additional authentication step in the process of users. This often happens at a place in the workflow where they don't expect it.
This is a situation that remain until the main login is adapted to the newer/higher security standard.
This RFC bridges the gap so the software can already be changed to accept newer protocols and the user can be made aware of upcoming changes without going through an authentication step using additional means.

## 2. Terminology

* **Client application**: The application that requires access.
* **Node**: A piece of software implementing the Nuts RFCs.
* **Resource server**: The application \(a protected resource\) that requires authorized access to its API’s.
* **VerifiableCredential**: Verifiable Credential according to the [Verifiable Credential Data Model](https://www.w3.org/TR/vc-data-model/).
* **VerifiablePresentation**: Verifiable Presentation according to the [Verifiable Credential Data Model](https://www.w3.org/TR/vc-data-model/).

## 3. Nuts Employee Identity

The authentication flow is as follows:

- a user requests access to a resource that is hosted at a *resource server* and requires additional authentication.

> - the *client application* starts a session on the *node* with the current user data.

Reviewer: Maybe coin the term "user data" (maybe change it to "user session info/data") as "the user data from the existing authenticated session in the client application", or something?

Author: "user data from the active session"

> - the user is directed to a web page where it accepts the given conditions by pressing a button.

Reviewer: or rejects

> - this button initiates a POST request to the *node for the given session.

Reviewer: No need for HTTP specifics here;

```suggestion
- this button sends the confirmation to the *node* for the given session.
```

> - upon receiving the request, the node will issue a `NutsEmployeeCredential` from and to the organization DID of the current user context.

Reviewer: ```suggestion
- upon receiving the request, the node will issue a `NutsEmployeeCredential` from and to the organization DID associated with the existing user session.
```

> - the credential is used to create a VerifiablePresentation.

Reviewer: a special kind, so name it here

- the VerifiablePresentation is used in the access token request as defined in [RFC003](./rfc003-oauth2-authorization.md#422-payload).

### 3.1 User data

In order for the resource server to comply to local legislation the following data MUST be added to the credential:

- **Initials**: The initials of the user.

> - **Familyname**: The family name of the user.

Reviewer: Are these references to the actual fields, or just descriptions?

```suggestion
- **Family name**: The family name of the user.
```

- **Identifier**: A unique identifier for the user within the context of the organization. This may be an email address or an employee number. 

> The user data MAY also contain a `RoleName` that contains the user role within the organization.

Reviewer: ```suggestion
The user data MAY also contain a `roleName` that contains the user role within the organization.
```

If using backticks here (so naming the exact field), do the same for the 3 above

### 3.2 HTML based challenge

> After a session has been started by the node, the client application MUST direct the user to a web page that is hosted by the node.

Reviewer: ```suggestion
After a session has been started by the node, the client application MUST present the user a web page that is hosted by the node.
```

The node MUST make sure that the web page is linked to the session and that the give user data can not be altered.

> The web page MUST present the user with a contract as stated in [RFC002](./rfc002-authentication-token.md#5-login-contract) and show which user data will be exposed to the resource server. 

Reviewer: The current UI says "Als u akkoord gaat worden deze gegevens gebruikt bij het opvragen van patient-gegevens bij andere zorginstellingen", so it's quite ambiguous what it will be used for (thus the UI isn't 100% according to this spec). Maybe it should be changed to "worden deze gegevens gedeeld met zorginstellingen"?

> The web page MUST also provide a checkbox that requires the user to acknowledge the challenge and confirm the data to be shared.

Reviewer: There's no checkbox in the UI, so maybe this section is a bit too much on the HOW instead of focussing on WHAT (e.g. "The web page must clearly indicate how the user can confirm and give consent, and how the user can reject the challenge").

> The web page MUST present a button that allows the user to submit the response to the node via an HTTP POST call.

Reviewer: too much detail I'd say (HTTP POST)
